From b15db38d7463e877aa98936160a5e21994b3d46a Mon Sep 17 00:00:00 2001
From: Vlad Yasevich <vyasevic@redhat.com>
Date: Fri, 29 Aug 2014 21:23:15 -0400
Subject: [net] macvlan: Initialize vlan_features to turn on offload support

Message-id: <1409347403-2401-5-git-send-email-vyasevic@redhat.com>
Patchwork-id: 89928
O-Subject: [RHEL7 PATCH 04/12] macvlan: Initialize vlan_features to turn on offload support.
Bugzilla: 1135619
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: John Linville <linville@redhat.com>

From: Vlad Yasevich <vyasevic@redhat.com>

Macvlan devices do not initialize vlan_features.  As a result,
any vlan devices configured on top of macvlans perform very poorly.
Initialize vlan_features based on the vlan features of the lower-level
device.

Signed-off-by: Vlad Yasevich <vyasevic@redhat.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 081e83a78db9b0ae1f5eabc2dedecc865f509b98)
Signed-off-by: Vladislav Yasevich <vyasevic@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/macvlan.c b/drivers/net/macvlan.c
index 7e868aa..b71dda3 100644
--- a/drivers/net/macvlan.c
+++ b/drivers/net/macvlan.c
@@ -501,6 +501,7 @@ static int macvlan_init(struct net_device *dev)
       (lowerdev->state & MACVLAN_STATE_MASK);
  dev->features   = lowerdev->features & MACVLAN_FEATURES;
  dev->features  |= ALWAYS_ON_FEATURES;
+ dev->vlan_features = lowerdev->vlan_features & MACVLAN_FEATURES;
  dev->gso_max_size = lowerdev->gso_max_size;
  dev->iflink  = lowerdev->ifindex;
  dev->hard_header_len = lowerdev->hard_header_len;
-- 
1.7.1